### ISSUE TITLE
Kubelet successfully registered node but it does not show up in API

### ISSUE THREAD

**Issue by lentzi90** [kubernetes-sigs/cluster-api#8631]
### What steps did you take and what happened?

I have created a Cluster with 1 control plane node and 1 worker using CAPI and CAPO.
When trying to upgrade from k8s v1.26.3 to v1.26.4 I followed these steps:

1. Create new OpenStackMachineTemplates for the new image
2. Switch the KCP and MD to use the new templates.

The KCP started to rollout, a new machine was created and provisioned, but the node never appeared in the API. Before debugging any deeper I decided to just delete the machine and see if there was some temporary issue. That broke the cluster. The API server started crashlooping and I had to ssh to the machines to be able to get information about what happened.

After some debugging I found out that the new control plane machine had managed to start its etcd pod and it had successfully joined the etcd cluster. So when I deleted the machine, it basically broke 1 of 2 etcd members and caused etcd to lose quorum. Since the machine was already gone I decided to restart the remaining (original) etcd container with `--force-new-cluster` to get rid of the extra member. With this everything started working again. Then another new machine was created to replace the one I deleted and the same thing happened again (except I didn't delete it this time).

I suspect it is some kind of configuration issue on my part because I have a hard time believing there is some bug in the node registration that didn't show up until now.

Here is a relevant snippet of kubelet logs:

```
May 10 09:26:40 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: I0510 09:26:40.760947    1046 kubelet_node_status.go:70] "Attempting to register node" node="prow-control-plane-v1-26-4-xlvqs"
May 10 09:26:40 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: E0510 09:26:40.763688    1046 kubelet_node_status.go:92] "Unable to register node with API server" err="nodes is forbidden: User \"system:anonymous\" cannot create resource \"nodes\" in API group \"\" at the cluster scope" node="prow-control-plane-v1-26-4-xlvqs"
May 10 09:26:41 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: W0510 09:26:41.793456    1046 reflector.go:424] vendor/k8s.io/client-go/informers/factory.go:150: failed to list *v1.Service: services is forbidden: User "system:anonymous" cannot list resource "services" in API group "" at the cluster scope
May 10 09:26:41 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: E0510 09:26:41.793524    1046 reflector.go:140] vendor/k8s.io/client-go/informers/factory.go:150: Failed to watch *v1.Service: failed to list *v1.Service: services is forbidden: User "system:anonymous" cannot list resource "services" in API group "" at the cluster scope
May 10 09:26:42 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: W0510 09:26:42.370279    1046 reflector.go:424] vendor/k8s.io/client-go/informers/factory.go:150: failed to list *v1.Node: nodes "prow-control-plane-v1-26-4-xlvqs" is forbidden: User "system:anonymous" cannot list resource "nodes" in API group "" at the cluster scope
May 10 09:26:42 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: E0510 09:26:42.370466    1046 reflector.go:140] vendor/k8s.io/client-go/informers/factory.go:150: Failed to watch *v1.Node: failed to list *v1.Node: nodes "prow-control-plane-v1-26-4-xlvqs" is forbidden: User "system:anonymous" cannot list resource "nodes" in API group "" at the cluster scope
May 10 09:26:42 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: W0510 09:26:42.880613    1046 reflector.go:424] vendor/k8s.io/client-go/informers/factory.go:150: failed to list *v1.RuntimeClass: runtimeclasses.node.k8s.io is forbidden: User "system:anonymous" cannot list resource "runtimeclasses" in API group "node.k8s.io" at the cluster scope
May 10 09:26:42 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: E0510 09:26:42.880713    1046 reflector.go:140] vendor/k8s.io/client-go/informers/factory.go:150: Failed to watch *v1.RuntimeClass: failed to list *v1.RuntimeClass: runtimeclasses.node.k8s.io is forbidden: User "system:anonymous" cannot list resource "runtimeclasses" in API group "node.k8s.io" at the cluster scope
May 10 09:26:43 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: W0510 09:26:43.241130    1046 reflector.go:424] vendor/k8s.io/client-go/informers/factory.go:150: failed to list *v1.CSIDriver: csidrivers.storage.k8s.io is forbidden: User "system:anonymous" cannot list resource "csidrivers" in API group "storage.k8s.io" at the cluster scope
May 10 09:26:43 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: E0510 09:26:43.241493    1046 reflector.go:140] vendor/k8s.io/client-go/informers/factory.go:150: Failed to watch *v1.CSIDriver: failed to list *v1.CSIDriver: csidrivers.storage.k8s.io is forbidden: User "system:anonymous" cannot list resource "csidrivers" in API group "storage.k8s.io" at the cluster scope
May 10 09:26:43 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: I0510 09:26:43.871104    1046 transport.go:135] "Certificate rotation detected, shutting down client connections to start using new credentials"
May 10 09:26:44 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: E0510 09:26:44.327962    1046 eviction_manager.go:261] "Eviction manager: failed to get summary stats" err="failed to get node info: node \"prow-control-plane-v1-26-4-xlvqs\" not found"
May 10 09:26:44 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: E0510 09:26:44.549344    1046 csi_plugin.go:295] Failed to initialize CSINode: error updating CSINode annotation: timed out waiting for the condition; caused by: nodes "prow-control-plane-v1-26-4-xlvqs" not found
May 10 09:26:45 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: E0510 09:26:45.574505    1046 csi_plugin.go:295] Failed to initialize CSINode: error updating CSINode annotation: timed out waiting for the condition; caused by: nodes "prow-control-plane-v1-26-4-xlvqs" not found
May 10 09:26:46 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: E0510 09:26:46.802326    1046 nodelease.go:49] "Failed to get node when trying to set owner ref to the node lease" err="nodes \"prow-control-plane-v1-26-4-xlvqs\" not found" node="prow-control-plane-v1-26-4-xlvqs"
May 10 09:26:47 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: E0510 09:26:47.175034    1046 csi_plugin.go:295] Failed to initialize CSINode: error updating CSINode annotation: timed out waiting for the condition; caused by: nodes "prow-control-plane-v1-26-4-xlvqs" not found
May 10 09:26:47 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: I0510 09:26:47.210274    1046 kubelet_node_status.go:70] "Attempting to register node" node="prow-control-plane-v1-26-4-xlvqs"
May 10 09:26:47 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: I0510 09:26:47.432080    1046 kubelet_node_status.go:73] "Successfully registered node" node="prow-control-plane-v1-26-4-xlvqs"
May 10 09:26:47 prow-control-plane-v1-26-4-xlvqs kubelet[1046]: E0510 09:26:47.472942    1046 kubelet_node_status.go:458] "Error getting the current node from lister" err="node \"prow-control-plane-v1-26-4-xlvqs\" not found"
```

Attaching more logs from the node that failed to join here:

[kubelet.log](https://github.com/kubernetes-sigs/cluster-api/files/11441900/kubelet.log)
[kube-apiserver.log](https://github.com/kubernetes-sigs/cluster-api/files/11441902/kube-apiserver.log)
[etcd.log](https://github.com/kubernetes-sigs/cluster-api/files/11441903/etcd.log)
[cloud-final.log](https://github.com/kubernetes-sigs/cluster-api/files/11441904/cloud-final.log)

### What did you expect to happen?

The new node should successfully join the cluster, or at least not "half-join" in this way that the etcd member was added but not the node itself.

### Cluster API version

v1.4.0

### Kubernetes version

```
❯ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.3", GitCommit:"9e644106593f3f4aa98f8a84b23db5fa378900bd", GitTreeState:"clean", BuildDate:"2023-03-15T13:40:17Z", GoVersion:"go1.19.7", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.3", GitCommit:"9e644106593f3f4aa98f8a84b23db5fa378900bd", GitTreeState:"clean", BuildDate:"2023-03-15T13:33:12Z", GoVersion:"go1.19.7", Compiler:"gc", Platform:"linux/amd64"}
```

### Anything else you would like to add?

There relevant part in the kubelet code: 
https://github.com/kubernetes/kubernetes/blob/626b2db1ca313723f0c39b894af631334ac02acc/pkg/kubelet/kubelet_node_status.go#L70-L75

### Label(s) to be applied

/kind bug
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

**Comment by lentzi90** [comment]
After some more debugging and a hunch I discovered that the names of the OpenStackMachineTemplates seemed to be the issue. I had included dots in them, like `prow-control-plane-v1.26.4`. Changing it to `prow-control-plane-v1-26-4` made it work.
Since then I have tried to reproduce this with CAPD but without success. Therefore I think this may be a bug in CAPO and this issue can be closed.

**Comment by killianmuldoon** [comment]
Thanks for following up on this. I think `.` is a disallowed character in many Kubernetes object types.
